### PR TITLE
feat: add samples for Datastore Admin

### PR DESCRIPTION
**Pull request by jdpedrie**
These tests take ~15 minutes to execute.

### CODE REVIEW

File: datastoreadmin/test/IndexTest.php [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+namespace Google\Cloud\Samples\DatastoreAdmin;
+
+use Google\Cloud\TestUtils\TestTrait;
+use PHPUnit\Framework\TestCase;
+
+class IndexTest extends TestCase
+{
+    use TestTrait {
+        TestTrait::runFunctionSnippet as traitRunFunctionSnippet;
```
Comment (bshaffer) [inline datastoreadmin/test/IndexTest.php:26]: No need for this override, just use `runFunctionSnippet`

File: datastoreadmin/test/IndexTest.php [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    public static function setUpBeforeClass(): void
+    {
+        self::checkProjectEnvVars();
```
Comment (bshaffer) [inline datastoreadmin/test/IndexTest.php:31]: This also shouldn't be necessary, it'll be called automatically using the `@beforeClass` annotation.

File: datastoreadmin/test/EntitiesTest.php [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+namespace Google\Cloud\Samples\DatastoreAdmin;
+
+use Google\Cloud\TestUtils\TestTrait;
+use PHPUnit\Framework\TestCase;
+
+class EntitiesTest extends TestCase
+{
+    use TestTrait {
+        TestTrait::runFunctionSnippet as traitRunFunctionSnippet;
```
Comment (bshaffer) [inline datastoreadmin/test/EntitiesTest.php:26]: Same here!

File: datastoreadmin/test/EntitiesTest.php [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    public static function setUpBeforeClass(): void
+    {
+        self::checkProjectEnvVars();
```
Comment (bshaffer) [inline datastoreadmin/test/EntitiesTest.php:31]: Same here!

File: datastoreadmin/src/index_get.php [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+// [START datastore_admin_index_get]
+use Google\Cloud\Datastore\Admin\V1\DatastoreAdminClient;
+
+/**
+ * Gets an index from Cloud Datastore
+ *
+ * @param string $projectId The Google Cloud project ID.
+ * @param string $indexId The resource ID of the index to get
+ */
+function index_get($projectId, $indexId) {
```
Comment (bshaffer) [inline datastoreadmin/src/index_get.php:35]: nit: put curly brace on its own line
```suggestion
function index_get($projectId, $indexId)
{
```

File: datastoreadmin/src/index_list.php [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+// [START datastore_admin_index_list]
+use Google\Cloud\Datastore\Admin\V1\DatastoreAdminClient;
+
+/**
+ * Lists the indexes that match the specified filters
+ *
+ * @param string $projectId The Google Cloud project ID.
+ * @param string $filter [optional] A filtering string.
+ */
+function index_list($projectId, $filter = "") {
```
Comment (bshaffer) [inline datastoreadmin/src/index_list.php:35]: Same here: 
```suggestion
function index_list($projectId, $filter = "")
{
```

File: datastoreadmin/src/index_delete.php [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+// [START datastore_admin_index_delete]
+use Google\Cloud\Datastore\Admin\V1\DatastoreAdminClient;
+
+/**
+ * Delete an index from Cloud Datastore
+ *
+ * @param string $projectId The Google Cloud project ID.
+ * @param string $indexId The ID of the index to delete.
+ */
+function index_delete($projectId, $indexId) {
```
Comment (bshaffer) [inline datastoreadmin/src/index_delete.php:35]:
```suggestion
function index_delete($projectId, $indexId)
{
```

File: datastoreadmin/src/index_create.php [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+use Google\Cloud\Datastore\Admin\V1\Index;
+use Google\Cloud\Datastore\Admin\V1\Index\AncestorMode;
+
+/**
+ * Create a Cloud Datastore index.
+ *
+ * @param string $projectId The Google Cloud project ID.
+ * @param string $kind The entity kind to which this index applies.
+ */
+function index_create($projectId, $kind) {
```
Comment (bshaffer) [inline datastoreadmin/src/index_create.php:37]:
```suggestion
function index_create($projectId, $kind)
{
```

File: datastoreadmin/src/entities_import.php [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+
+/**
+ * Imports entities into Google Cloud Datastore
+ *
+ * @param string $projectId The Google Cloud project ID.
+ * @param string $inputUri The full resource URL of the external storage
+ *     location. Currently only Cloud Storage is supported. Values should be of
+ *     form `gs://bucket-name/folder-name`.
+ */
+function entities_import($projectId, $inputUri) {
```
Comment (bshaffer) [inline datastoreadmin/src/entities_import.php:37]:
```suggestion
function entities_import($projectId, $inputUri)
{
```

File: datastoreadmin/src/entities_export.php [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+
+/**
+ * Exports entities from Google Cloud Datastore
+ *
+ * @param string $projectId The Google Cloud project ID.
+ * @param string $outputUrlPrefix Location for the export metadata and data
+ *     files. Currently only Cloud Storage is supported. Values should be of
+ *     form `gs://bucket-name/folder-name`.
+ */
+function entities_export($projectId, $outputUrlPrefix) {
```
Comment (bshaffer) [inline datastoreadmin/src/entities_export.php:37]:
```suggestion
function entities_export($projectId, $outputUrlPrefix)
{
```

File: datastoreadmin/quickstart.php [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+
+// [START datastore_admin_client_create]
+# Includes the autoloader for libraries installed with composer
+require __DIR__ . '/vendor/autoload.php';
+
+# Imports the Datastore Admin client library
+use Google\Cloud\Datastore\Admin\V1\DatastoreAdminClient;
+
+# Your Google Cloud Platform project ID
+$projectId = 'YOUR_PROJECT_ID';
```
Comment (bshaffer) [inline datastoreadmin/quickstart.php:26]: this is never used. Did you mean to pass this into `DatastoreAdminClient`?

File: datastoreadmin/src/entities_export.php [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+    $admin = new DatastoreAdminClient();
+
+    $operation = $admin->exportEntities($projectId, $outputUrlPrefix);
+
+    $operation->pollUntilComplete();
+    if ($operation->operationSucceeded()) {
+        printf("The export operation succeeded. File location is %s", $operation->getResult()->getOutputUrl());
```
Comment (bshaffer) [inline datastoreadmin/src/entities_export.php:44]: nit, but all of these `printf` statements should use single quotes and add newlines!

```suggestion
        printf('The export operation succeeded. File location is %s' . PHP_EOL, $operation->getResult()->getOutputUrl());
```

**Review by bshaffer — CHANGES_REQUESTED** [review]

**Comment by jdpedrie** [comment]
@bshaffer these tests are reliably passing locally against multiple projects, and failing reliably on kokoro against multiple projects. Is there perhaps something I'm missing with the configuration of the kokoro projects?

File: datastoreadmin/test/EntitiesTest.php [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+    public static function setUpBeforeClass(): void
+    {
+        self::requireEnv('GOOGLE_STORAGE_BUCKET');
+    }
+
+    public function testEntitiesLifecycle()
+    {
+        $objectName = uniqid("datastoreadmin-export-");
+        $uri = uniqid(sprintf(
+            "gs://%s/%s",
```
Comment (bshaffer) [inline datastoreadmin/test/EntitiesTest.php:37]:
```suggestion
            "gs://%s-datastoreadmin/%s",
```

File: datastoreadmin/src/entities_export.php [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+ * Exports entities from Google Cloud Datastore
+ *
+ * @param string $projectId The Google Cloud project ID.
+ * @param string $outputUrlPrefix Location for the export metadata and data
+ *     files. Currently only Cloud Storage is supported. Values should be of
+ *     form `gs://bucket-name/folder-name`.
+ */
+function entities_export($projectId, $outputUrlPrefix)
+{
+    $admin = new DatastoreAdminClient();
```
Comment (bshaffer) [inline datastoreadmin/src/entities_export.php:39]: We should explicitly set the `projectId` whenever it's being passed in, to make sure the project ID of the client and the project ID of the operation are the same. 

```suggestion
    $admin = new DatastoreAdminClient([
        'projectId' => $projectId,
    ]);
```

**Comment by bshaffer** [comment]
I believe the cause of the problem is when tested on projects using "Datastore Legacy", the operation to create the index pass, but when tested on projects using "Firestore in Datastore Mode", the tests fail.

**Comment by meredithslota** [comment]
My understanding is that the PHP Datastore client has an autogenerated admin client library (source: @kolea2) and thus has autogenerated code samples to go with it. Those code samples are found here: https://github.com/googleapis/google-cloud-php/tree/main/DatastoreAdmin/samples/V1/DatastoreAdminClient Closing this accordingly to reduce duplication.

**Pull request closed without merging**